File: log-format-enforcer-maven-plugin/src/it/java-it/src/test/java/com/leandronunes85/lfe/maven_plugin/it/java/LogFormatEnforcerTest.java
package com.leandronunes85.lfe.maven_plugin.it.java;

import com.leandronunes85.lfe.maven_plugin.it.java.LogFormatEnforcer;
import org.junit.Test;

import java.util.ArrayList;


public class LogFormatEnforcerTest {

    private final LogFormatEnforcer victim = LogFormatEnforcer.loggerFor(LogFormatEnforcerTest.class);

    @Test
    public void basicSyntaxTest() {
        victim.info(messageBuilder -> messageBuilder.mandatory1("valueMandatory1").mandatory2(100)
                .optional1(new ArrayList()).optional2("value")
                .and("other", "otherValue").exception(new RuntimeException()));
    }

    private String costlyOperation() {
        throw new RuntimeException("This method should not have been called!");
    }

    public void logValuesAreNotComputedIfTheLogIsNotPerformed() {
        victim.trace(messageBuilder -> messageBuilder
                .mandatory1("valueMandatory1").mandatory2(costlyOperation())
        );
    }
}

File: log-format-enforcer-maven-plugin/src/it/kotlin1.3-it/src/test/kotlin/com/leandronunes85/lfe/maven_plugin/it/kotlin/LogFormatEnforcerTest.kt
package com.leandronunes85.lfe.maven_plugin.it.kotlin

import com.leandronunes85.lfe.maven_plugin.it.kotlin.LogFormatEnforcer
import com.leandronunes85.lfe.maven_plugin.it.kotlin.LogFormatEnforcerExtensions.traceEach
import com.leandronunes85.lfe.maven_plugin.it.kotlin.LogFormatEnforcerExtensions.debug

import org.junit.Test
import java.lang.Thread


class LogFormatEnforcerTest {

    private val victim = LogFormatEnforcer.loggerFor<LogFormatEnforcerTest>()


    @Test
    fun `basic syntax test`() {
        victim.info {
            mandatory1("").mandatory2(10)
                .optional1(emptyArray<Any>()).optional2("value")
                .and("other", "").exception(RuntimeException())
        }
    }

    @Test
    fun `extension function syntax`() {
        emptyArray<Int>()
                .traceEach(victim) { number: Int -> mandatory1(number).mandatory2("number is a single element") }
                .debug(victim) { wholeArray: Array<Int> -> mandatory1("").mandatory2("").and("wholeArray", wholeArray) }
    }

    @Test
    fun `log values are not computed if the log is not performed`() {
        fun costlyOperation(): String {
            throw RuntimeException("This method should not have been called!")
        }

        victim.trace {
            mandatory1("").mandatory2(costlyOperation())
        }
    }
}